src/databricks/labs/ucx/azure/resources.py

```diff
@@ -121,44 +115,149 @@ def inner() -> dict[str, str]:
 
         return _credentials
 
+    def get(self, path: str, api_version: str | None = None):
+        headers = {"Accept": "application/json"}
+        query = {}
+        if api_version is not None:
+            query = {"api-version": api_version}
+        return self._api_client.do("GET", path, query, headers)
+
+    def put(self, path: str, api_version: str | None = None, body: dict | None = None):
+        headers = {"Content-Type": "application/json"}
+        query: dict[str, str] = {}
+        if api_version is not None:
+            query = {"api-version": api_version}
+        if body is not None:
+            return self._api_client.do("PUT", path, query, headers, body)
+        return None
+
+    def post(self, path: str, body: dict | None = None):
+        headers = {"Content-Type": "application/json"}
+        query: dict[str, str] = {}
+        if body is not None:
+            return self._api_client.do("POST", path, query, headers, body)
+        return self._api_client.do("POST", path, query, headers)
+
+    def delete(self, path: str):
+        # this method is added only to be used in int test to delete the application once tests pass
+        headers = {"Content-Type": "application/json"}
+        query: dict[str, str] = {}
+        return self._api_client.do("DELETE", path, query, headers)
+
+    @property
+    def token(self):
+        return self._token_source.token()
+
+
+class AzureResources:
+    def __init__(self, azure_mgmt: AzureAPIClient, azure_graph: AzureAPIClient, include_subscriptions=None):
+        if not include_subscriptions:
+            include_subscriptions = []
+        self._azure_mgmt = azure_mgmt
+        self._azure_graph = azure_graph
+        self._include_subscriptions = include_subscriptions
+        self._role_definitions = {}  # type: dict[str, str]
+        self._principals: dict[str, Principal | None] = {}
+
     def _get_subscriptions(self) -> Iterable[AzureSubscription]:
-        for subscription in self._get_resource("/subscriptions", api_version="2022-12-01").get("value", []):
+        for subscription in self._azure_mgmt.get("/subscriptions", "2022-12-01").get("value", []):
             yield AzureSubscription(
                 name=subscription["displayName"],
                 subscription_id=subscription["subscriptionId"],
                 tenant_id=subscription["tenantId"],
             )
 
-    def _tenant_id(self):
-        token = self._token_source.token()
+    def create_service_principal(self) -> Principal:
+        try:
+            application_info: dict[str, str] = self._azure_graph.post(
+                "/v1.0/applications", {"displayName": "UCXServicePrincipal"}
```

Comment:
Bug: Use dynamic name and prompt the user for it. Use "unity-catalog-migration-{inventory_database}-{workspace_id}" as default.